### PR TITLE
feat: Crash event and session in same envelope

### PR DESCRIPTION
**Pull request by philipphofmann**
## :scroll: Description

Send the hard crash event in the same envelope as the crashed session so the Release Health
statistics show proper numbers.

This is a different way to solve the same problem as #716. Either we merge this PR or the other one.

## :bulb: Motivation and Context

We want to have proper Release Health statistics.

## :green_heart: How did you test it?
SentrySessionGeneratorTests and with smoke testing on an iPhone.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the CHANGELOG
- [x] No breaking changes

### CODE REVIEW

File: Sources/Sentry/SentryFileManager.m [hunk 1]
```diff
@@ -217,44 +225,72 @@ - (void)handleEnvelopesLimit
...
+- (void)storeSession:(SentrySession *)session sessionFilePath:(NSString *)sessionFilePath
 {
     NSData *sessionData = [SentrySerialization dataWithSession:session error:nil];
-    [SentryLog logWithMessage:[NSString stringWithFormat:@"Writing to current session: %@",
-                                        self.currentSessionFilePath]
+    [SentryLog logWithMessage:[NSString stringWithFormat:@"Writing session: %@", sessionFilePath]
                      andLevel:kSentryLogLevelDebug];
-    @synchronized(self.currentSessionFilePath) {
-        [sessionData writeToFile:self.currentSessionFilePath options:NSDataWritingAtomic error:nil];
+    @synchronized(_sessionLock) {
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryFileManager.m:242]: L: We could use `sessionFilePath` as the lock object here instead. Writing two different files don't need to synchronize. Assuming `sessionFilePath` is always the same insrtances (these are fields in the class (always same reference), correct?, And they can't be null, which would anyway break here when writing.

This is true to all usages of `_sessionLock` in this file.

File: Sources/Sentry/SentryFileManager.m [hunk 2]
```diff
@@ -217,44 +225,72 @@ - (void)handleEnvelopesLimit
...
-        currentSessionData = [fileManager contentsAtPath:self.currentSessionFilePath];
-        if (nil == currentSessionData) {
+    NSData *currentData = nil;
+    @synchronized(_sessionLock) {
+        currentData = [fileManager contentsAtPath:sessionFilePath];
+        if (nil == currentData) {
             return nil;
         }
-        SentrySession *currentSession = [SentrySerialization sessionWithData:currentSessionData];
+        SentrySession *currentSession = [SentrySerialization sessionWithData:currentData];
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryFileManager.m:289]: L: You could end the sync block before this, serialization can happen outside of the lock.

File: Sources/Sentry/SentryHub.m [hunk 3]
```diff
@@ -143,9 +137,9 @@ - (void)closeCachedSessionWithTimestamp:(NSDate *_Nullable)timestamp
                              andLevel:kSentryLogLevelDebug];
             [session endSessionExitedWithTimestamp:timestamp];
         }
+        [self deleteCurrentSession];
+        [client captureSession:session];
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryHub.m:141]: L: We could have a else block here logging, to help us see if things round in a different order during debugging.

**Review by bruno-garcia — APPROVED** [review]
A couple of improvements but LGTM! let's :shipit: ?

File: Sources/Sentry/SentryFileManager.m (hunk 1 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryFileManager.m:242]: Good, point. Thanks

File: Sources/Sentry/SentryFileManager.m (hunk 2 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryFileManager.m:286]: Yep makes sense. Was like this, good catch 💯

File: Sources/Sentry/SentryHub.m (hunk 3 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryHub.m:141]: I don't fully understand your point here sorry. Can you maybe explain it in a different way?

**Pull request merged**